### PR TITLE
chore: Chore add VNU GitHub Action

### PR DESCRIPTION
**Pull request by nschonni**
As and alternate to mdn/yari#418 
Downside to this approach is that there is no local runnable version of the testing without running the replacement that touches ~400 files

### CODE REVIEW

**Comment by nschonni** [comment]
@Elchi3 FYI, the issue it's failing on was introduced in #1106

**Comment by Elchi3** [comment]
@nschonni not sure I follow. What did I break? :) Can you elaborate?

**Comment by Elchi3** [comment]
ahh I see it now.

> Error: "file:/home/runner/work/content/content/files/en-us/web/html/global_attributes/enterkeyhint/index.html":23.112-23.115: error: No “p” element in scope but a “p” end tag seen.

Yeah, it would be great if CI or any tests would have caught this for me. I'll try to see if I can add an HTML validator to my editor at least.

**Comment by Elchi3** [comment]
Opened https://github.com/mdn/content/pull/1210 which should fix it.

**Comment by peterbe** [comment]
I'm not sure. This feels like a hack. And adding `vnu-jar` feels very heavy. Also, this kind of stuff feels more appropriate to do with Yari if anything. https://github.com/mdn/content/runs/1686915136 was cool though. 
There's just something that doesn't feel great about running the content through a HTML validator. It's not really HTML. It's Kumascript-HTML if anything. Perhaps it would make more sense to do a string replace on the `rawHTML` and replace it with something like `<ignore data-macro="..."/>` and then run it. 
Besides, the whole `sed` stuff feels weird and if we're going to do this properly, I think we should do it properly and with more care and time. Sorry for sounding a bit of a downer. I'm also hesitant to invest more into the HTML validation when we could use the limited time and energy investing into something Markdownish.

**Comment by sideshowbarker** [comment]
>  I'm also hesitant to invest more into the HTML validation when we could use the limited time and energy investing into something Markdownish.

Strongly agree. (And that coming from a developer of an HTML validator…) Feels like this might be a bit of an XY problem https://en.wikipedia.org/wiki/XY_problem.

It’d be great if a plan and proposed schedule for moving to something Markdown-ish were put together.

That said, if/when we do move to something Markdown-ish, does that means we necessarily mass-convert all the existing sources? Or instead would it be that — at least in the near term — we’d continue HTML file in the tree, mixed with some Markdown files?

**Comment by peterbe** [comment]
> > I'm also hesitant to invest more into the HTML validation when we could use the limited time and energy investing into something Markdownish.
> 
> Strongly agree. (And that coming from a developer of an HTML validator…) Feels like this might be a bit of an XY problem https://en.wikipedia.org/wiki/XY_problem.
> 

Never heard that one before. Thanks!

> It’d be great if a plan and proposed schedule for moving to something Markdown-ish were put together.
> 

I hear ya, but that's going to be hard. Sorry. The reason is that we predict it'll be a slow progress since we first need to do a bunch of exploration, research, and prototyping. Within that work, we'd get a better handle on compromises and trade-offs that we need to counter for the fact that we have so much legacy and weird content. 

> That said, if/when we do move to something Markdown-ish, does that means we necessarily mass-convert all the existing sources? Or instead would it be that — at least in the near term — we’d continue HTML file in the tree, mixed with some Markdown files?

I suspect we won't. It'd probably be gentle and almost a page-by-page. Or sub-tree by sub-tree.

**Comment by Ryuno-Ki** [comment]
> That said, if/when we do move to something Markdown-ish, does that means we necessarily mass-convert all the existing sources?

I feel like this should be moved into another issue, which could be linked to to collect all the comments that contain bits of what would be needed.

**Comment by nschonni** [comment]
@peterbe this continues to find issues as I rebase this occasionally. To make this more of a CI only thing, I could drop the package install and just "NPX" run it on the CI only

**Comment by nschonni** [comment]
@sideshowbarker I think we'll need a new NPM release for the SVG failure that I think you already fixed upstream

**Comment by nschonni** [comment]
@peterbe this continues to catch stuff. I've suppressed the other issue pending another vnu-jar release. Can you please land this?

**Comment by sideshowbarker** [comment]
I strongly support landing this — because I can say from experience that it’s caught bugs/mistakes I’ve personally made, and from looking at other commits, I see that in general it’s regularly catching bugs/mistakes that we’d otherwise not be catching.

I recognize we have a plan to eventually move everything to Markdown. But that is relatively a long way away still — so in the meantime, for as long as we continue to have an HTML sources in the tree (prior to whenever we do managed to get everything fully moved to Markdown) this is serving a very useful purpose. So it would be great to see it get landed.

**Comment by sideshowbarker** [comment]
> @sideshowbarker I think we'll need a new NPM release for the SVG failure that I think you already fixed upstream

I just now published a new vnu-jar npm — version 21.4.9

**Comment by nschonni** [comment]
> I just now published a new vnu-jar npm — version 21.4.9 (sorry for not having gotten to it before now)

No problem, I've bumped to the latest, and removed the SVG error message suppression

**Comment by sideshowbarker** [comment]
https://github.com/mdn/content/pull/4205 is yet another example of this catching a markup mistake I personally made. My point being that I doubt I’m the only one who’s gonna end up introducing new markup errors like that one from time to time — and so that reaffirms my feeling that it would be great to see this change get landed.

**Comment by peterbe** [comment]
You know, we could consider using the PR Review Companion to turn the HTML validation outputs into nice warnings. 
Installing vnu-jar on a server is a very different story that "forcing" everyone to have a valid and working Java runtime. 

Also, if there are outputs that shouldn't make the CI go red (fail), the Review Companion is a much better alternative.

**Comment by peterbe** [comment]
> @peterbe so I've tried it here with the "pr companion" approach, and it's not going to work as it can't properly report the annotations back to the PR. Please take a look before I revert back to the previous approach.

To be honest, I don't understand it. 
I know we've talked about this for many days but I still haven't had a chance to look at any of that. 
I still have basic stuff like https://github.com/mdn/yari/issues/3487 on my maybe-to-do list. :)

> I've you've got a concern with adding the install, I can do it CI-only, but there is nothing requiring people to have java installed, unless they want to run the task. I think it's better to keep the task in the package.json, so the folks that do want to run it locally, can. Those that don't would just install one extra npm package that they don't invoke

"one extra npm package" ? 
How would that work?

**Comment by nschonni** [comment]
OK, I've rolled back to the original version again. I'll try to summarized what happens here.

- `vnu-jar` is added to the dependencies, and a script is added to the package.json, so people can invoke the same thing that the CI will run. Contributors aren't required to run this locally, and the NPM package is the JAR file and a wrapper script. This isn't adding much to the their local install process, and they don't need to have Java, unless they want to run this extra check locally, rather than on CI
- When a PR is opened, that touches an HTML file (or the supporting config files), the new job will run
- This is an optional job (like all the CI jobs), so a maintainer can always land a PR, even if the job is red. HTML issues should likely get fixed as part of the PR review, rather than ignored 😄 
- This CI job registers a handler and adds in-line comments on the files tab (like below). This approach doesn't conflict with your PR companion, they're separate concerns, and we can continue the discussion on the other issue
image

**Comment by peterbe** [comment]
> OK, I've rolled back to the original version again. I'll try to summarized what happens here.
> 
>     * `vnu-jar` is added to the dependencies, and a script is added to the package.json, so people can invoke the same thing that the CI will run. Contributors aren't required to run this locally, and the NPM package is the JAR file and a wrapper script. This isn't adding much to the their local install process, and they don't need to have Java, unless they want to run this extra check locally, rather than on CI
> 

So like that it's perfectly fine. If it's harmless and cheap to install the wrapper script without it complaining if you don't have `java` installed, then I think it's OK for everyone to have to install it even if they never run it. 

>     * When a PR is opened, that touches an HTML file (or the supporting config files), the new job will run
> 

How about making it part of the regular Yari runner? 
We can have the default be off but in `pr-test.yml` we have something like:
```
  env:
    - BUILD_RUN_VNU_VALIDATOR=true
```
It could even be part of the flaw system even though it's run on the source HTML file, not the built HTML. 

>     * This is an optional job (like all the CI jobs), so a maintainer can always land a PR, even if the job is red. HTML issues should likely get fixed as part of the PR review, rather than ignored 😄
> 

I'd rather the CI doesn't become red ❌ unless there's important things that really ought to be fix now. 
Some things *can't* be fixed if you get a ❌ such as an attempt to edit the system files. 

>     * This CI job registers a handler and adds in-line comments on the files tab (like below). This approach doesn't conflict with your PR companion, they're separate concerns, and we can continue the discussion on the other issue
>       image

So you got it working?? You had said: *"and it's not going to work as it can't properly report the annotations back to the PR."*

**Comment by nschonni** [comment]
> How about making it part of the regular Yari runner?

I don't think the complication of trying to add it to Yari is worth it. This job will go away at the end of your markdown conversion. This just keeps the files in better shape for editors (and the markdown conversion tool) till you eventually cut over.

> It could even be part of the flaw system even though it's run on the source HTML file, not the built HTML.

Think @sideshowbarker mentioned before that this works best in "batch mode" like we're calling it here, vs individual invokations that each take a Java start-up hit.


> I'd rather the CI doesn't become red ❌ unless there's important things that really ought to be fix now.

HTML validation issues are important, as either they creating rendering issues, or the browser is swallowing the issue, but may stop doing after an unrelated change on the page

> So you got it working?? You had said: _"and it's not going to work as it can't properly report the annotations back to the PR."_

It's always been working as this current separate PR job. It didn't work when I tried to move it to a `workflow` trigger as part of the PR Companion. It's possible that way is workable with annotations, but you're fighting the built-in integration that Actions already has with the problem matchers. The implications for the other PR companion stuff is tracked in https://github.com/mdn/yari/issues/3488, and probably better to discuss there

**Comment by nschonni** [comment]
Sorry, this has been going for awhile (started in another PR), so I forgot the "gotcha" for someone running locally, is that they'd get some "false positives" from the elements in the frontmatter on some pages.
CI modifies the local tree in place to escape that, but I didn't think it was "safe" to add that to the locally NPM script, so people don't accidentally change a large chunk of files.

**Comment by nschonni** [comment]
@peterbe ping

**Comment by peterbe** [comment]
> @peterbe ping

So how do we test this? My fear is that false positives (where VNU incorrectly complains about a file the user didn't even touch) leads to a PR being failingly red when the user didn't introduce the problem. Perhaps they just touched a file that's always had an HTML validation problem. 

Will `yarn vnu-jar` exit non-zero if there's a `index.html` that has a validation problem? Or will it just post a comment within the PR?

Perhaps butter me up and/or we put in some things to work around possible false positives and then we merge this and see how it goes.

File: package.json [hunk 1]
```diff
@@ -7,6 +7,7 @@
     "node": ">=12.0.0"
   },
   "scripts": {
+    "vnu-jar": "java -jar node_modules/vnu-jar/build/dist/vnu.jar --errors-only --filterfile .vnurc --no-langdetect --skip-non-html files/",
```
Comment (peterbe) [inline package.json:10]: How does it know which files to test? Will it test all 11k each time?

Comment (nschonni) [reply package.json:10]: Tests them all in a batch each time, but it's around 3 minutes in the CI. User doesn't have to run it locally unless they want to

**Comment by nschonni** [comment]
> My fear is that false positives (where VNU incorrectly complains about a file the user didn't even touch) leads to a PR being failingly red when the user didn't introduce the problem.

If the error is with the file that is touched by the PR, it will show up on the regular files as inline annotations. If it's an issue with something they didn't touch, then GitHub pushes them to the bottom of the Files tab.
Either way, as we tested with the "System file changes" job, a contributor can always land even with the job failing if they don't want to bother a new contributor 😄

File: package.json (hunk 1 above)
Comment (peterbe) [reply package.json:10]: Seems here https://github.com/mdn/content/pull/1206/checks?check_run_id=2493264692 it only took 1m 32s. 
23 seconds was the time it took to run the actual `yarn vnu-jar`. 

Will this delay when the PR Review Companion starts to run? I know that one waits for the `PR Test` workflow.

**Comment by peterbe** [comment]
If the time it takes to run this is too slow, I don't want the reviewer (with merge access) to have to sit around impatiently waiting for the "HTML Linting" to go green whilst it's checking unrelated files. If that's the case we can optimize this workflow to just test the `files/**/index.html` files that were actually touched.

**Comment by Ryuno-Ki** [comment]
If I understand that correctly, the failing test requires admin permissions to resolve?

**Comment by peterbe** [comment]
> If I understand that correctly, the failing test requires admin permissions to resolve?

No, just write access.

**Comment by peterbe** [comment]
Let's try it! We'll keep an eye on it to see what things it's finding and what might not be great and we can iterate from there.

**Comment by nschonni** [comment]
> If I understand that correctly, the failing test requires admin permissions to resolve?

PS: @Ryuno-Ki we did some testing around that in https://github.com/mdn/content/pull/3728

File: package.json (hunk 1 above)
Comment (nschonni) [reply package.json:10]: Nope, this runs in parallel with the other jobs AFAIK. PR companion is waiting for the "PR Test" to finish, not all other parallel jobs

**Comment by peterbe** [comment]
@nschonni To be honest, I don't know how to test this. 
I check out your branch and run 
```
yarn vnu-jar
```
I get pages and pages of warnings and then it exits non-zero.

Screen Shot 2021-05-03 at 4 59 18 PM

How come it's not failing here in this PR?

**Comment by nschonni** [comment]
> How come it's not failing here in this PR?

Right, the CI is doing this to escape the front matter https://github.com/mdn/content/pull/1206/files#diff-c6f76479cefbaa8eea1ca9824164c77d44d8b981fbd44ae665e36cb09e6951f2R45

**Comment by peterbe** [comment]
> > How come it's not failing here in this PR?
> 
> Right, the CI is doing this to escape the front matter https://github.com/mdn/content/pull/1206/files#diff-c6f76479cefbaa8eea1ca9824164c77d44d8b981fbd44ae665e36cb09e6951f2R45

Oh so it rewrites the `index.html` files first. Then you'd get different results.

**Comment by nschonni** [comment]
It only re-writes the frontmatter lines in the CI, so the lines/colums it reports here are still correct

**Comment by nschonni** [comment]
@peterbe https://github.com/mdn/content/pull/4901 fixes the latest round of issues this flagged so this can be landed before/after that one

**Comment by nschonni** [comment]
@peterbe and there were more between landing the latest and rebasing 😄 
I'll submit a cleanup PR after if you land this

**Comment by peterbe** [comment]
> @peterbe and there were more between landing the latest and rebasing 😄
> I'll submit a cleanup PR after if you land this

I did see the PR you submitted recently. Great catches! That kind of stuff would have been prevented with this as part of CI. Right?

**Comment by nschonni** [comment]
Yup, that's what this catches, but doesn't block a PR from landing if a maintainer is OK with cleaning it up after

**Comment by nschonni** [comment]
@peterbe any remaining concerns?

**Comment by peterbe** [comment]
> @peterbe any remaining concerns?

Can you merge in main one more time to assert that all errors will go away. 

Also, I don't understand your point about being able to merge even if the line CI fails. Won't that just ruin CI for the next person making a PR if you merge it.

**Comment by nschonni** [comment]
@peterbe rebased again
Yes the next persons CI might go red, but I've said I'd submit another PR to clean it up (like the 31 others I submitted while this has been open https://github.com/mdn/content/pulls?q=is%3Apr+html+validation+is%3Aclosed+author%3Anschonni)

**Comment by peterbe** [comment]
I don't like...

* This runs for every file and not just the ones touched. 
* The complicated `sed` command doesn't work on any OS. 
* That this whole thing relies on changing the file which makes it problematic to run locally

https://github.com/nschonni/mdn-content/pull/1 solves 1 and 2 of that list. 

What's so weird is that if someone, a reviewer, says "Fine. I'll let this pass" it'll poison every other PR going forward. 
It also just feels strange to run something like `vnu-jar` on all ~12k files. It could maybe make sense in a `push` workflow or as a cronjob.

**Comment by peterbe** [comment]
> [nschonni#1](https://github.com/nschonni/mdn-content/pull/1) solves 1 and 2 of that list.

Perhaps I rushed that. The front-matter doesn't actually need to escaped. It just needs to be deleted, right?
Using that `bodyBegin` attribute from `fm()` we could turn...
```html
---
foo: "bar"
---
<p>Here it starts</p>
```
...into...
```html



<p>Here it starts</p>
```
the line number still make sense in the problem matcher and front-matter could never matter.

**Comment by nschonni** [comment]
OK, I give up. @sideshowbarker I'll leave this to you to work out internally

**Comment by peterbe** [comment]
> OK, I give up. @sideshowbarker I'll leave this to you to work out internally

Please don't. If you agree and can incorporate my PR I think this is good to go.

File: package.json [hunk 2]
```diff
@@ -7,6 +7,7 @@
     "node": ">=12.0.0"
   },
```
Comment (lisafabian19) [inline package.json:8]: libareyrefund1/2reciptsreturnlocationSAHC177mr.modelC177sportsrecoverylocationBaoxian2hiCXI83747choclategrowDVRSAFC177//.SAF17CFR#134.117.158.83growthrateexhangeatcorrectionsfacityandreleasepod2018-2025releasesmakenonknownsecretadeplantes2fabianGirard,correctionresearchadd20thousandforreleasetoplantforresourcestothesecretdevelopmentforvideoandmedia-200thosuandisoverpopulatedsouthpacificincludingamerica>

**Review by lisafabian19 — COMMENTED** [review]
> 

> hurryup

**Pull request closed without merging**